Use payload from triggerHookWithToken in code-review-testing hook

La0 opened the pull request:
Refs https://github.com/mozilla/code-review/issues/2518

The goal of that PR is to support payloads sent by Phabricator webhooks using `triggerHookWithToken` Taskcluster API call, without altering the other triggers (direct & from pulse).

The payload sent by Phabricator is described on [that bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1935142#c6):

```json
{
  "object": {
    "type": "DREV",
    "phid": "PHID-DREV-cljdgxaey4cxvxb4knjp"
  },
...
}
```

I only need the object type & phid as environment variables in the created task for now.

jcristau commented on hooks/project-relman/code-review-testing.yml:
```diff
@@ -31,6 +31,13 @@ payload:
         else: {}
         then:
           $eval: payload
+      - $if: firedBy == 'triggerHookWithToken'
```
Should this be something like `$if: payload`, instead of depending on the mechanism?

La0 replied:
We should only get these variables when the hook is triggered through a token (as this covers a different workflow than the current one).

I can add extra conditions to make sure we have the variables set, but I think it would remain clearer to have the `firedBy` condition

jcristau replied:
Fine (plus this is following the existing pattern, which I still think is not great but not worth changing now).

jcristau reviewed (changes requested):
We should also update the trigger_schema in hooks.yml to match the expectations here.

La0 commented:
I rebased the branch and described the trigger schema in `hooks.yml`

Could you have another look @jcristau :pray:

La0 commented:
I fixed the typos, and added support for `payload.transactions` as [dkl mentionned on Bugzilla I'll need them](https://bugzilla.mozilla.org/show_bug.cgi?id=1935142#c11).

I'm unsure about the schema payload, as this is my first time writing these...

I tested the json-e expression on [the playground](https://json-e.js.org/playground.html) using:

Template:
```yaml
  env:
    $merge:
      - $if: firedBy == 'triggerHook'
        else: {}
        then:
          $eval: payload
      - $if: firedBy == 'triggerHookWithToken'
        else: {}
        then:
          $let:
            transactions: 
             $map: {$eval: payload.transactions}
             each(transaction): {$eval: transaction.phid}
          in:
           PHABRICATOR_OBJECT_TYPE:
             $eval: payload.object.type
           PHABRICATOR_OBJECT_PHID:
             $eval: payload.object.phid
           TRANSACTIONS: {$json: {$eval: transactions}}

      - $if: firedBy == 'pulseMessage'
        else: {}
        then:
          TRY_RUN_ID:
            $eval: payload.runId
          TRY_TASK_GROUP_ID:
            $eval: payload.status.taskGroupId
          TRY_TASK_ID:
            $eval: payload.status.taskId
```

Context (as YAML converted from [payload provided by dkl](https://bugzilla.mozilla.org/show_bug.cgi?id=1935142#c6)):

```yaml
firedBy: 'triggerHookWithToken'
payload:
 object:
  type: DREV
  phid: PHID-DREV-cljdgxaey4cxvxb4knjp
 triggers:
  - phid: PHID-HRUL-y4g3jm6qrqarufzfohfq
 action:
  test: false
  silent: false
  secure: true
  epoch: 1733442143
 transactions:
  - phid: PHID-XACT-DREV-uxsvjvneur37nue
  - phid: PHID-XACT-DREV-qulb5hkuij2ugz6
  - phid: PHID-XACT-DREV-hsygmkod3scvfxo
  - phid: PHID-XACT-DREV-6vhe4gqkpmhlhvy
  - phid: PHID-XACT-DREV-hk5zrs54cw7lr3w
  - phid: PHID-XACT-DREV-ublcnlitaq5sh3x
  - phid: PHID-XACT-DREV-tyevaf5wulxmn7y
  - phid: PHID-XACT-DREV-nmx7mfwyao7iul2
  - phid: PHID-XACT-DREV-pal453i7g2iglpq
  - phid: PHID-XACT-DREV-62sdvz3h2ii4jkj
  - phid: PHID-XACT-DREV-bw346binj5r4v6h
  - phid: PHID-XACT-DREV-pnra5n5toc7cdz4

```

which results in :

```yaml
env:
  PHABRICATOR_OBJECT_TYPE: DREV
  PHABRICATOR_OBJECT_PHID: PHID-DREV-cljdgxaey4cxvxb4knjp
  TRANSACTIONS: >-
    ["PHID-XACT-DREV-uxsvjvneur37nue","PHID-XACT-DREV-qulb5hkuij2ugz6","PHID-XACT-DREV-hsygmkod3scvfxo","PHID-XACT-DREV-6vhe4gqkpmhlhvy","PHID-XACT-DREV-hk5zrs54cw7lr3w","PHID-XACT-DREV-ublcnlitaq5sh3x","PHID-XACT-DREV-tyevaf5wulxmn7y","PHID-XACT-DREV-nmx7mfwyao7iul2","PHID-XACT-DREV-pal453i7g2iglpq","PHID-XACT-DREV-62sdvz3h2ii4jkj","PHID-XACT-DREV-bw346binj5r4v6h","PHID-XACT-DREV-pnra5n5toc7cdz4"]
```

jcristau reviewed (approved):
Thanks!

La0 commented:
Thanks a lot for the merge @jcristau , but I see there has been a failure on [CI deployment](https://github.com/mozilla-releng/fxci-config/actions/runs/12256070783/job/34190714113) whereas the CI check was OK.

Is there a way to troubleshoot the schema syntax before merging ?

jcristau commented:
Not sure.  In this case the problem is `int` instead of `integer` for runId, as far as I can tell.
